hooks might affect the internal fields slice of a child logger

Hooks are designed to be able to modify the provided fields slice. But this slice might also contain some fields from a child logger created using `WithFields` function. The problem is that modifying these fields might affect the internal slice of a child logger, which is meant to be static. So, multiple hook calls might cause unexpected behaviour like this:

```go
// adds _ prefix to each key.
prefixHook := func(lvl log.Level, msg string, fields []log.Field) error {
	for i := range fields {
		fields[i].Key = "_" + fields[i].Key
	}
	return nil
}

logger = log.WithHooks(logger, prefixHook)
logger = log.WithFields(logger, log.String("key", "value"))

logger.Error("call 1") // prints [ERROR] call 1 _key=value
logger.Error("call 2") // prints [ERROR] call 2 __key=value
logger.Error("call 3") // prints [ERROR] call 3 ___key=value
```

The reason is that two slices might share the same array, so any changes are visible from both sides. The fix is to copy the internal slice before appending it to the rest of the provided fields.

--- a/log.go
+++ b/log.go
@@ -87,15 +87,15 @@ type withFields struct {
 }
 
 func (wf *withFields) Debug(msg string, fields ...Field) {
-	wf.logger.Debug(msg, append(wf.fields, fields...)...)
+	wf.logger.Debug(msg, append(wf.copyFields(), fields...)...)
 }
 
 func (wf *withFields) Info(msg string, fields ...Field) {
-	wf.logger.Info(msg, append(wf.fields, fields...)...)
+	wf.logger.Info(msg, append(wf.copyFields(), fields...)...)
 }
 
 func (wf *withFields) Error(msg string, fields ...Field) {
-	wf.logger.Error(msg, append(wf.fields, fields...)...)
+	wf.logger.Error(msg, append(wf.copyFields(), fields...)...)
 }
 
 func (wf *withFields) AddCallerSkip(skip int) {
@@ -105,6 +105,12 @@ func (wf *withFields) AddCallerSkip(skip int) {
 }
 
 func (wf *withFields) Unwrap() Logger { return wf.logger }
+
+func (wf *withFields) copyFields() []Field {
+	fields := make([]Field, len(wf.fields))
+	copy(fields, wf.fields)
+	return fields
+}
 
 // Level indicates a logging priority.
 type Level int

--- a/log_test.go
+++ b/log_test.go
@@ -13,8 +13,8 @@ import (
 )
 
 func TestWithFields(t *testing.T) {
-	var testLogger testLogger
-	logger := log.WithFields(&testLogger, log.Int("foo", 1))
+	var spy spyLogger
+	logger := log.WithFields(&spy, log.Int("foo", 1))
 	logger.Info("first call", log.Int("bar", 2))
 	logger.Info("second call", log.Int("baz", 3))
 
@@ -28,7 +28,7 @@ func TestWithFields(t *testing.T) {
 			fields: []log.Field{log.Int("foo", 1), log.Int("baz", 3), log.String("caller", "log_test.go:19")},
 		},
 	}
-	if got := testLogger.calls; !reflect.DeepEqual(got, want) {
+	if got := spy.calls; !reflect.DeepEqual(got, want) {
 		t.Errorf("got %+v; want %+v", got, want)
 	}
 }
@@ -66,8 +66,8 @@ func TestWithHooks(t *testing.T) {
 		}
 	}
 
-	var testLogger testLogger
-	logger := log.WithHooks(&testLogger, prefixHook, multiplierHook, eofHook)
+	var spy spyLogger
+	logger := log.WithHooks(&spy, prefixHook, multiplierHook, eofHook)
 
 	logger.Debug("first call", log.Int("foo", 1))
 	logger.Info("second call", log.Int("bar", 2))
@@ -82,7 +82,44 @@ func TestWithHooks(t *testing.T) {
 			fields: []log.Field{log.Int("_bar", 4), log.String("caller", "log_test.go:73")},
 		},
 	}
-	if got := testLogger.calls; !reflect.DeepEqual(got, want) {
+	if got := spy.calls; !reflect.DeepEqual(got, want) {
+		t.Errorf("got %+v; want %+v", got, want)
+	}
+}
+
+// https://github.com/junk1tm/log/issues/11
+func TestIssue11(t *testing.T) {
+	// adds _ prefix to each key.
+	prefixHook := func(lvl log.Level, msg string, fields []log.Field) error {
+		for i := range fields {
+			fields[i].Key = "_" + fields[i].Key
+		}
+		return nil
+	}
+
+	var spy spyLogger
+	logger := log.WithHooks(&spy, prefixHook)
+	logger = log.WithFields(logger, log.String("key", "value"))
+
+	logger.Debug("first call")
+	logger.Info("second call")
+	logger.Error("third call")
+
+	want := []call{
+		{
+			msg:    "first call",
+			fields: []log.Field{log.String("_key", "value"), log.String("caller", "log_test.go:104")},
+		},
+		{
+			msg:    "second call",
+			fields: []log.Field{log.String("_key", "value"), log.String("caller", "log_test.go:105")},
+		},
+		{
+			msg:    "third call",
+			fields: []log.Field{log.String("_key", "value"), log.String("caller", "log_test.go:106")},
+		},
+	}
+	if got := spy.calls; !reflect.DeepEqual(got, want) {
 		t.Errorf("got %+v; want %+v", got, want)
 	}
 }
@@ -92,29 +129,30 @@ type call struct {
 	fields []log.Field
 }
 
-type testLogger struct {
+// spyLogger records its calls for later inspection in tests.
+type spyLogger struct {
 	calls      []call
 	callerSkip int
 }
 
-func (tl *testLogger) Debug(msg string, fields ...log.Field) {
-	tl.calls = append(tl.calls, call{msg: msg, fields: append(fields, tl.callerField())})
+func (sl *spyLogger) Debug(msg string, fields ...log.Field) {
+	sl.calls = append(sl.calls, call{msg: msg, fields: append(fields, sl.callerField())})
 }
 
-func (tl *testLogger) Info(msg string, fields ...log.Field) {
-	tl.calls = append(tl.calls, call{msg: msg, fields: append(fields, tl.callerField())})
+func (sl *spyLogger) Info(msg string, fields ...log.Field) {
+	sl.calls = append(sl.calls, call{msg: msg, fields: append(fields, sl.callerField())})
 }
 
-func (tl *testLogger) Error(msg string, fields ...log.Field) {
-	tl.calls = append(tl.calls, call{msg: msg, fields: append(fields, tl.callerField())})
+func (sl *spyLogger) Error(msg string, fields ...log.Field) {
+	sl.calls = append(sl.calls, call{msg: msg, fields: append(fields, sl.callerField())})
 }
 
-func (tl *testLogger) AddCallerSkip(skip int) {
-	tl.callerSkip += skip
+func (sl *spyLogger) AddCallerSkip(skip int) {
+	sl.callerSkip += skip
 }
 
-func (tl *testLogger) callerField() log.Field {
-	_, file, line, _ := runtime.Caller(tl.callerSkip + 2)
+func (sl *spyLogger) callerField() log.Field {
+	_, file, line, _ := runtime.Caller(sl.callerSkip + 2)
 	file = file[strings.LastIndex(file, "/")+1:]
 	value := fmt.Sprintf("%s:%d", file, line)
 

